### PR TITLE
MB-31210: Support Fuzzy MatchPhrase, MultiPhrase and Phrase queries

### PR DESCRIPTION
**Pull request by CascadingRadium**
## Jira

[MB-31210](https://issues.couchbase.com/browse/MB-31210)

## Description
Added fuzziness support for MatchPhrase queries.

### CODE REVIEW

File: search/searcher/search_phrase_test.go
Comment (metonymic-smokey) [inline search/searcher/search_phrase_test.go:133]: I think adding a new UT to demonstrate fuzziness would be good.

Comment (CascadingRadium) [reply search/searcher/search_phrase_test.go:133]: UT?

File: search/query/multi_phrase.go
Comment (abhinavdangeti) [inline search/query/multi_phrase.go:33]: I'm not sure prefix makes much sense with multi phrase, because the order of terms with multi phrase query does not matter. Also this is more in line with `TermQuery` so let's keep the definition of the MultiPhraseQuery like it was before and apply these new attributes only for a `MatchPhraseQuery`.

Comment (CascadingRadium) [reply search/query/multi_phrase.go:33]: But a match phrase query / phrase query is internally a multi-phrase query, so it must also be supported.
A prefix essentially is applied to the fuzzy search of each term in the Phrase, so it works out like so, each word will be fuzzily matched to a word in the index, provided that a prefix for the word is present.
The unit test file also reflects the same.

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]
Recommending a code change here.

File: search/query/multi_phrase.go
Comment (CascadingRadium) [reply search/query/multi_phrase.go:33]: basically yeah, lets say i have a match phrase for 
"Apple banana" it becomes a multi phrase query for
[["apple"]["banana"]], and for both of these terms, we have a fuzzy search being done,
So FuzzySearch ("apple", Fuzziness, prefix) is done, so both fuzziness+prefix must be passed from the top level searcher.

If i have MultiPhraseQuery of:
[[ "apple"   ],[ "banana" ] ]
      "red"           "yellow"

Q1 = DisjunctionQuery of FuzzyQueries over "apple" and "red"
Q2 = DisjunctionQuery of FuzzyQueries over "banana" and "yellow"
And result Q = ConjQuery(Q1,Q2). So i have to send prefix via the top searcher or hard code it to 0 while allowing for a user defined fuzziness

Comment (CascadingRadium) [reply search/query/multi_phrase.go:33]: In Elastic search, Fuzzy Match phrase is not supported out of the box, but can be implemented using some workaround as suggested in this post:
https://stackoverflow.com/questions/38816955/elasticsearch-fuzzy-phrases

Basically use a span query that allows fine-tuning - **for each word the user can give a fuzziness and a prefix length** there as in this image 
image

The below image shows a match provided a per word fuzziness and prefix
image


So what i can suggest is - We allow this restricted match phrase query which keeps the fuzziness and prefix same throughout, and then add a functionality to implement span queries for more fine tuning.

Comment (CascadingRadium) [reply search/query/multi_phrase.go]: Done

**Review by abhinavdangeti — APPROVED** [review]
Looks fine to me. 
@CascadingRadium Let's wait for one more review possibly.

**Pull request merged**
